src/entity_resolver/core/graph.py

```diff
@@ -1,44 +1,188 @@
+        else:
+            names = value.split()
+            first = names[-1]
+            if len(names) == 1:
+                last = ''
+            else:
+                last = ''.join(names[:-1]).strip()
+            return [first, last]
+
+
 
 
 class Node:
 
-    def __init__(self, node_id, attrs):
+    def __init__(self, node_id, edge, attrs):
+        """
+        :param node_id: int
+        :param edge: edge object
+        :param attrs:  list of Attribute()
+        """
         self.id = node_id
-        self.attrs = attrs
-        pass
+        self.edge = edge
+        self.attrs = {}
+        for attr in attrs:
+            self.attrs[attr.name] = attr
 
     def __hash__(self):
         return self.id
 
     def __eq__(self, other):
         return self.id == other.id
 
+    def get_attr_names(self):
+        '''
+        :return: list of attribute names
+        '''
+        return list(self.attrs.keys())
 
-class Edge:
+    def get_attr(self, name):
+        '''
+        :param name: name of attribute
+        :return: list of tokenized words if atttribute type is string
+        '''
+        attr = self.attrs[name]
+        return attr.value
 
-    def __init__(self, edge_id, nodes, attrs):
+
+class Edge:
+    def __init__(self, edge_id, attrs=None):
+        """
+        :param edge_id: int
+        :param attrs: list of Attribute()
+        """
         self.id = edge_id
-        self.nodes = nodes
+        self.nodes = []
         self.attrs = attrs
-        pass
+
+    def add_node(self, node):
+        """
+        :param node: Node() object
+        """
+        self.nodes.append(node)
 
 
 class Graph:
 
-    def __init__(self, nodes, edges):
-        self.nodes = nodes
-        self.edges = edges
-        pass
+    def __init__(self, nodes, edges, attr_types):
+        """
+        :param nodes: iterables (list) of Node()
+        :param edges: dictionary of edges with key edge_id and value Edge()
+        """
+        self.nodes = list(nodes)
+        self.edges = list(edges.values())
+        self.edge_dict = edges
+        self.attr_types = attr_types
+        self.id2node = {}
+        self.id2edge = {}
+        for node in self.nodes:
+            self.id2node[node.id] = node
+        for edge in self.edges:
+            self.id2edge[edge.id] = edge
+
+    def add_nodes(self, new_nodes):
+        """
+        :param new_nodes:  iterables (list) of Node()
+        """
+        for node in new_nodes:
+            self.id2node[node.id] = node
+        self.nodes.extend(new_nodes)
+
+    def add_edges(self, new_edges):
+        """
+        :param new_edges:  iterables (list) of Edge()
+        """
+        for edge in new_edges:
+            self.id2edge[edge.id] = edge
+        self.nodes.extend(new_edges)
+
+    def get_neighbors(self, node):
+        """
+        :param node object
+        :return: list of Node() that are neighbors of node_id
+        """
+        return self.edge_dict[node.edge].nodes
```

Comment:
By storing the associated `Edge` object in the `Node` object, one may simply use `node.edge.nodes`